File: src/domain/entities/column.py
from pathlib import Path
from typing import Optional
from pydantic import BaseModel, Field
from src.core.types import ColumnId, ParentId, Timestamp, FilePath
from src.utils.string_utils import generate_id_from_name
from src.utils.date_utils import now
from src.domain.entities.item import Item


class Column(BaseModel):
    id: ColumnId = Field(default="")
    name: str
    position: int = 0
    limit: Optional[int] = None
    created_at: Timestamp = Field(default_factory=now)
    items: list[Item] = Field(default_factory=list)
    file_path: Optional[FilePath] = None

    def model_post_init(self, __context) -> None:
        if self.file_path and not self.id:
            dir_name = Path(self.file_path).parent.name
            self.id = dir_name
            if not self.name or self.name == dir_name:
                self.name = dir_name.replace("-", " ").replace("_", " ").title()
        elif not self.id:
            self.id = generate_id_from_name(self.name) or "unnamed_column"

    def update(self, **kwargs) -> None:
        for key, value in kwargs.items():
            if hasattr(self, key):
                setattr(self, key, value)

    def add_item(self, title: str, parent_id: Optional[ParentId] = None, item_id: Optional[str] = None) -> Item:
        """Add a new item to the column.

        Args:
            title: Item title
            parent_id: Optional parent ID for grouping
            item_id: Optional pre-generated item ID (used by service layer)

        Returns:
            The created Item
        """
        item = Item(id=item_id or "", title=title, parent_id=parent_id, column_id=self.id)
        self.items.append(item)
        return item

    def move_item_to_end(self, item: Item) -> bool:
        item.move_to_column(self.id)
        self.items.append(item)
        return True

    def remove_item(self, item_id: str) -> bool:
        original_count = len(self.items)
        self.items = [item for item in self.items if item.id != item_id]
        return len(self.items) < original_count

    def get_all_items(self) -> list[Item]:
        return self.items

    def get_item_by_id(self, item_id: str) -> Optional[Item]:
        for item in self.items:
            if item.id == item_id:
                return item
        return None

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "position": self.position,
            "limit": self.limit,
            "created_at": self.created_at,
        }

File: mobile/src/domain/entities/Column.ts
/**
 * Column entity for organizing items in a kanban board
 * Ported from Python: src/domain/entities/column.py
 */

import { ColumnId, ParentId, Timestamp, FilePath } from "../../core/types";
import { now } from "../../utils/dateUtils";
import { generateIdFromName } from "../../utils/stringUtils";
import { Item } from "./Item";

export interface ColumnProps {
  id?: ColumnId;
  name: string;
  position?: number;
  limit?: number | null;
  created_at?: Timestamp;
  items?: Item[];
  file_path?: FilePath | null;
}

export class Column {
  id: ColumnId;
  name: string;
  position: number;
  limit: number | null;
  created_at: Timestamp;
  items: Item[];
  file_path: FilePath | null;

  constructor(props: ColumnProps) {
    this.name = props.name;
    this.position = props.position || 0;
    this.limit = props.limit !== undefined ? props.limit : null;
    this.created_at = props.created_at || now();
    this.items = props.items || [];
    this.file_path = props.file_path !== undefined ? props.file_path : null;

    // Auto-generate ID if not provided
    if (props.id) {
      this.id = props.id;
    } else if (this.file_path) {
      // Extract ID from file path (directory name)
      const pathParts = this.file_path.split("/");
      const dirName = pathParts[pathParts.length - 2] || pathParts[pathParts.length - 1];
      this.id = dirName;
      if (!this.name || this.name === dirName) {
        this.name = dirName.replace(/-/g, " ").replace(/_/g, " ");
        this.name = this.name
          .split(" ")
          .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
          .join(" ");
      }
    } else {
      this.id = generateIdFromName(this.name) || "unnamed_column";
    }
  }

  /**
   * Update column properties
   */
  update(updates: Partial<ColumnProps>): void {
    Object.assign(this, updates);
  }

  /**
   * Add a new item to the column
   */
  addItem(title: string, parentId?: ParentId | null, itemId?: string): Item {
    const item = new Item({
      id: itemId || "",
      title,
      parent_id: parentId,
      column_id: this.id,
    });
    this.items.push(item);
    return item;
  }

  /**
   * Move an existing item to the end of this column
   */
  moveItemToEnd(item: Item): boolean {
    item.moveToColumn(this.id);
    this.items.push(item);
    return true;
  }

  /**
   * Remove an item from the column
   */
  removeItem(itemId: string): boolean {
    const originalCount = this.items.length;
    this.items = this.items.filter((item) => item.id !== itemId);
    return this.items.length < originalCount;
  }

  /**
   * Get all items in the column
   */
  getAllItems(): Item[] {
    return this.items;
  }

  /**
   * Find an item by ID
   */
  getItemById(itemId: string): Item | null {
    return this.items.find((item) => item.id === itemId) || null;
  }

  /**
   * Convert to plain object for serialization
   */
  toDict(): Record<string, any> {
    return {
      id: this.id,
      name: this.name,
      position: this.position,
      limit: this.limit,
      created_at: this.created_at,
    };
  }

  /**
   * Create Column from plain object (deserialization)
   */
  static fromDict(data: Record<string, any>): Column {
    return new Column({
      id: data.id,
      name: data.name,
      position: data.position,
      limit: data.limit,
      created_at: data.created_at ? new Date(data.created_at) : undefined,
      file_path: data.file_path,
    });
  }
}
